Feat: use framework info

erezrokah opened the pull request:
### Summary

Fixes https://github.com/netlify/cli/issues/1495
Fixes https://github.com/netlify/cli/issues/1252

- [x] Verify this works with https://github.com/netlify/cli/pull/2312

Related to https://github.com/netlify/cli/pull/843

### Changes in behavior

Changes from #843

- `framework = "#custom"`. 
  - Current behavior - an error is shown if `command` and `targetPort` are not configured. 
  - New behavior - same as current.
- `framework` is not `#custom` and user has specified `command` and `targetPort`. 
  - Current behavior - an error is shown to set `framework = "#custom"`. 
  - New behavior - everything works as the CLI has all the information it needs to run. This allows running `ntl dev --command "echo hello" --targetPort 8000` which was not supported before (adding `--framework "#custom"` didn't work as well).
 - `dir` flag is specified.
    - Current behavior - we don't run framework detection, use a static server and show an error if `targetPort` or `command` are specified.
    - New behavior - we don't run framework detection and use a static server. 
      - If `command` is configured we print a warning and suggest to use `targetPort` for a custom setup.
      - If `targetPort` is configured we print a warning and suggest to use `staticServerPort` to configure the static server port.

Other changes
- `framework = "create-react-app"` (or any other specific framework)
  - Current behavior - run the `create-react-app` detector, [if it fails report an error](https://github.com/netlify/cli/blob/8152bb838177ee7c2666c6cfad8b9f0523fbe293/src/utils/detect-server.js#L135), otherwise use it
  - New behavior - always use the specific framework settings

- ~~Added a new `--framework` flag. You can only configure `framework` using the `dev` block at the moment. That doesn't make sense since we already have `command` and `targetPort` flags.~~ Done in https://github.com/netlify/cli/pull/2818

### Test plan

~~Testing this manually at the moment.~~
~~Will think of a good way to test the integration without having a performance hit.~~

~~Tested only the happy path, so others are probably broken.~~

~~Replaced the unit tests with integration tests~~

Tests were added in https://github.com/netlify/cli/pull/2821, I've updated them to reflect the new behavior

**- A picture of a cute animal (not mandatory but encouraged)**
![image](https://user-images.githubusercontent.com/26760571/103928685-df2ef080-5124-11eb-9751-d7cc16030a88.png)

erezrokah commented on src/commands/dev/index.js:
```diff
@@ -53,14 +53,14 @@ const isNonExistingCommandError = ({ command, error }) => {
 
   // we use reject=false to avoid rejecting synchronously when the command doesn't exist
-  const frameworkProcess = execa(settings.command, settings.args, { preferLocal: true, reject: false })
+  const frameworkProcess = execa.command(settings.command, { preferLocal: true, reject: false })
```
`framework-info` returns a single string of the command + arguments

erezrokah commented on src/utils/detect-server-settings.js:
```diff
@@ -0,0 +1,322 @@
+  }
+}
+
+const detectServerSettings = async (devConfig, flags, projectDir, log) => {
```
This is main function, the other ones are utility functions

erezrokah commented on src/utils/detect-server-settings.js:
```diff
@@ -0,0 +1,322 @@
+const { EOL } = require('os')
```
Renamed from `detect-server.js`

erezrokah commented on tests/framework-detection.test.js:
```diff
@@ -65,7 +65,7 @@ test('should use static server when framework is set to #static', async (t) => {
 })
 
-test('should report error if using static server and `command` is configured', async (t) => {
+test('should warn if using static server and `command` is configured', async (t) => {
```
Expected change in behavior

erezrokah commented on tests/framework-detection.test.js:
```diff
@@ -74,18 +74,19 @@ test('should report error if using static server and `command` is configured', a
 })
 
-test('should error if using static server and `targetPort` is configured', async (t) => {
+test('should warn if using static server and `targetPort` is configured', async (t) => {
```
Expected change in behavior

erezrokah commented on tests/framework-detection.test.js:
```diff
@@ -94,14 +95,19 @@ test('should error if using static server and `targetPort` is configured', async
 })
 
-test('should error when `command` and `targetPort` when configured, but framework is not', async (t) => {
+test('should run `command` when both `command` and `targetPort` are configured', async (t) => {
```
Expected change in behavior

erezrokah commented on tests/framework-detection.test.js:
```diff
@@ -117,10 +123,11 @@ test('should error when `command` and `targetPort` when configured, but framewor
 })
 
-test('should error when a specific framework is configured but not detected', async (t) => {
+test('should force a specific framework when configured', async (t) => {
```
Expected change in behavior

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -9,14 +9,13 @@ Generated by [AVA](https://avajs.dev).
     ◈ Running static server from "site-with-index-file"␊
     ␊
-    ◈ Server listening to 88888␊
+    ◈ Static server listening to 88888␊
```
Improvement to the log

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -31,10 +30,10 @@ Generated by [AVA](https://avajs.dev).
     ◈ Running static server from "site-with-index-file/public"␊
     ␊
-    ◈ Server listening to 88888␊
+    ◈ Static server listening to 88888␊
```
Improvement to the log

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -49,14 +48,13 @@ Generated by [AVA](https://avajs.dev).
-    ◈ No app server detected and no "command" specified␊
-    ◈ Using current working directory␊
-    ◈ Unable to determine public folder to serve files from␊
+    ◈ Using simple static server because '[dev.framework]' was set to '#static'␊
```
The previous log message was less accurate.

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
     `◈ Netlify Dev ◈␊
-    ◈ Using simple static server because --dir flag was specified␊
+    ◈ Using simple static server because '--dir' flag was specified␊
+    ◈ Ignoring command setting since using a simple static server. Configure 'command' and 'targetPort' for a custom setup␊
```
`Ignoring command...` is the new warning instead of failure

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
     `◈ Netlify Dev ◈␊
-    ◈ Using simple static server because --dir flag was specified␊
+    ◈ Using simple static server because '--dir' flag was specified␊
+    ◈ Ignoring 'targetPort' setting since using a simple static server.␊
```
`Ignoring 'targetPort'...` is the new warning instead of failure

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
     `◈ Netlify Dev ◈␊
-    ◈ "framework" option must be set to "#custom" when specifying both "command" and "targetPort" options`
+    ◈ Starting Netlify Dev with custom config␊
+    hello␊
```
Expected change in behavior

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
 
     `◈ Netlify Dev ◈␊
-    ◈ Specified "framework" detector "create-react-app" did not pass requirements for your project`
+    ◈ Starting Netlify Dev with Create React App␊
```
Expected change in behavior, if the user configures `framework = 'create-react-app'` we simply use it instead of running that detector and failing if it doesn't match

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
 
     `◈ Netlify Dev ◈␊
-    ◈ Unsupported value provided for "framework" option in config. Please use "#custom" if you're using a framework not intrinsically supported by Netlify Dev. E.g. with "command" and "targetPort" options. Or use one of following values: "angular", "brunch", "create-react-app", "docusaurus", "eleventy", "ember", "expo", "gatsby", "gridsome", "hexo", "hugo", "jekyll", "middleman", "next", "nuxt",  ...
+    ◈ Invalid framework "to-infinity-and-beyond-js". It should be one of: docusaurus, docusaurus-v2, eleventy, gatsby, gridsome, hexo, hugo, jekyll, middleman, next-nx, next, blitz, nuxt, phenomic, react-static, redwoodjs, stencil, vuepress, assemble, docpad, harp, metalsmith, roots, wintersmith, cecil, zola, angular, create-react-app, ember, expo, quasar, quasar-v0.17, sapper, svelte, svelte-kit ...
```
Should be improved by https://github.com/netlify/framework-info/pull/342

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
 
     `◈ Netlify Dev ◈␊
-    ◈ Starting Netlify Dev with create-react-app␊
+    ◈ Starting Netlify Dev with Create React App␊
```
This is better as we print framework names instead of ids

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -66,77 +64,100 @@ Generated by [AVA](https://avajs.dev).
 > Snapshot 1
 
     `◈ Netlify Dev ◈␊
-    ◈ Overriding command with setting derived from netlify.toml [dev] block: echo hello␊
```
This log message is wrong. The tests doesn't set the command via `[dev]` block, but via a flag:
https://github.com/netlify/cli/blob/08d87ff6b4785612ddf97e54098ced40d58bffb5/tests/framework-detection.test.js#L180

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -146,7 +167,9 @@ Generated by [AVA](https://avajs.dev).
 > Snapshot 1
 
     `◈ Netlify Dev ◈␊
-    ◈ Overriding command with setting derived from netlify.toml [dev] block: oops-i-did-it-again forgot-to-use-a-valid-command␊
```
Same wrong log message:
https://github.com/netlify/cli/blob/08d87ff6b4785612ddf97e54098ced40d58bffb5/tests/framework-detection.test.js#L199

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -146,7 +167,9 @@ Generated by [AVA](https://avajs.dev).
 
     `◈ Netlify Dev ◈␊
-    ◈ Overriding command with setting derived from netlify.toml [dev] block: oops-i-did-it-again forgot-to-use-a-valid-command␊
+    ◈ Unable to determine public folder to serve files from. Using current working directory␊
```
The `Unable to determine public folder` message was missing in several flows

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -9,14 +9,13 @@ Generated by [AVA](https://avajs.dev).
+    ◈ No app server detected. Using simple static server␊
+    ◈ Unable to determine public folder to serve files from. Using current working directory␊
     ◈ Setup a netlify.toml file with a [dev] section to specify your dev server settings.␊
     ◈ See docs at: https://cli.netlify.com/netlify-dev#project-detection␊
```
I still need to go over `https://cli.netlify.com/netlify-dev` to make sure it makes sense

erezrokah commented on tests/snapshots/framework-detection.test.js.md:
```diff
@@ -156,16 +179,14 @@ Generated by [AVA](https://avajs.dev).
-      [gatsby] npm develop [3A[23D[41C[3B[2K[1A[2K[1A[2K[1A[2K[G? Multiple possible start commands found gatsby-start[53D[53C␊
-    Add \`framework = "gatsby"\` to [dev] section of your netlify.toml to avoid this selection prompt next time␊
-    ◈ Starting Netlify Dev with gatsby␊
+    > [Gatsby] 'npm run develop' ␊
```
This is better as we use framework names instead of ids

JGAntunes reviewed (approved):
🚀 thanks for the comments @erezrokah, super helpful 👍

erezrokah commented on docs/commands/dev.md:
```diff
@@ -40,8 +40,8 @@ netlify dev
-$ netlify dev -c "yarn start"
-$ netlify dev -c hugo
+$ netlify dev -d public
+$ netlify dev -c "hugo server -w" --targetPort 1313
```
Both `-c` and `targetPort` are required (this behavior didn't change)

erezrokah commented on docs/netlify-dev.md:
```diff
@@ -68,22 +68,21 @@ USAGE
-  -o, --offline              disables any features that require network access
-  -p, --port=port            Specify port of netlify dev
-  -l, --live                 Start a public live session
+  -c, --command=command                                        command to run
```
Aligned this with the current output of `ntl dev --help` and removed less important flags

erezrokah commented on docs/netlify-dev.md:
```diff
@@ -96,8 +95,6 @@ netlify dev --live
 
 You will get a URL that looks like `https://clever-cray-2aa156-6639f3.netlify.live/`. This can be accessed by anyone as long as you keep your session open.
 
-> Note: there are currently known issues with ending the live session alongside your webdevserver, as well as with live reloading. We are working on fixing it, and would appreciate repro cases, or you may check [existing issues with the `--live` tag](https://github.com/netlify/netlify-dev-plugin/issues?q=is%3Aissue+is%3Aopen+sort%3Aupdated-desc+label%3A--live). In the mean time you can run `ps au ...
```
This is outdated

erezrokah commented on docs/netlify-dev.md:
```diff
@@ -170,7 +166,7 @@ As for which port to use while doing local development in Netlify Dev, always se
    ┌──────────────────────────────────────────────────────────────┐
    │                                                              │
-   │   [Netlify Dev] Server now ready on http://localhost:64939   │
+   │   [Netlify Dev] Server now ready on http://localhost:8888    │
```
`8888` is the default port and what most users will experience

erezrokah commented:
@JGAntunes, added a bug fix in https://github.com/netlify/cli/pull/1704/commits/0a5a296bc682e2dc662d2da8bdaa9738d67100a1 and also updated the docs in https://github.com/netlify/cli/pull/1704/commits/878fe43f84ef2cb7d9fd6ba2a9267d10c1ff8cee
